velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -57,6 +81,128 @@ void ensureRegexIsCompatible(
   }
 }
 
+// Blocks patterns that contain character class union, intersection, or
+// difference because these are not understood by RE2 and will be parsed as a
+// different pattern than in java.util.regex.
+void ensureRegexIsConstantAndCompatible(
+    const char* functionName,
+    const VectorPtr& patternVector) {
+  if (!patternVector || !patternVector->isConstantEncoding()) {
+    VELOX_USER_FAIL("{} requires a constant pattern.", functionName);
+  }
+  if (patternVector->isNullAt(0)) {
+    return;
+  }
+  const StringView pattern =
+      patternVector->as<ConstantVector<StringView>>()->valueAt(0);
+  checkForCompatiblePattern(
+      std::string(pattern.data(), pattern.size()), functionName);
+}
+
+// REGEX_REPLACE(string, pattern, overwrite) → string
+// REGEX_REPLACE(string, pattern, overwrite, position) → string
+//
+// If a string has a substring that matches the given pattern, replace
+// the match in the string wither overwrite and return the string. If
+// optional paramter position is provided, only make replacements
+// after that positon in the string (1 indexed).
+//
+// If position <= 0, throw error.
+// If position > length string, return string.
+template <typename T>
+struct RegexReplaceFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace) {
+    re2::RE2* patternRegex = getCachedRegex(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
+
+    std::string string(stringInput.data(), stringInput.size());
+    RE2::GlobalReplace(&string, *patternRegex, replaceStringPiece);
+
+    if (string.size()) {
+      result.resize(string.size());
+      std::memcpy(result.data(), string.data(), string.size());
+    } else {
+      result.resize(0);
+    }
+  }
+
+  void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace,
+      const arg_type<int64_t>& position) {
+    VELOX_USER_CHECK_GE(position, 1, "REGEX_REPLACE requires a position >= 1");
+
+    re2::RE2* patternRegex = getCachedRegex(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
+    re2::StringPiece inputStringPiece = toStringPiece(stringInput);
+
+    // Adjust the position for UTF-8 by counting the code points.
+    size_t utf8Position = 0;
+    size_t codePointsCounted = 0;
+    while (codePointsCounted < position - 1 &&
+           utf8Position < stringInput.size()) {
+      if ((inputStringPiece.data()[utf8Position] & 0xc0) != 0x80) {
+        ++codePointsCounted;
+      }
+      ++utf8Position;
+    }
+
+    if (utf8Position >= stringInput.size()) {
```

Comment:
In this case, we do not need the following logic including creation of RE2::GlobalReplace. We can just set result to stringInput, no?

Reply:
Correct, I will implement this change tomorrow morning.

Reply:
@mbasmanova I implemented this logic but had to change the if statement clause to "utf8Position > stringInput.size()" to handle the case covered in regexReplaceEmptyStringsAndPatterns where we run regex_replace("", "", "prefix") -> "prefix".

Please let me know if you see anything else you would like me to change.